File: HTCommander.Core/WavFileWriter.cs
/*
Copyright 2026 Ylian Saint-Hilaire
Licensed under the Apache License, Version 2.0 (the "License");
http://www.apache.org/licenses/LICENSE-2.0
*/

using System;
using System.IO;

namespace HTCommander
{
    /// <summary>
    /// Simple cross-platform WAV file writer. Replaces NAudio.Wave.WaveFileWriter
    /// for recording PCM audio to disk. Supports 16-bit PCM only.
    /// </summary>
    public class WavFileWriter : IDisposable
    {
        private readonly FileStream _stream;
        private readonly BinaryWriter _writer;
        private readonly int _sampleRate;
        private readonly int _bitsPerSample;
        private readonly int _channels;
        private long _dataChunkSizePosition;
        private int _dataSize = 0;
        private bool _disposed = false;

        public WavFileWriter(string path, int sampleRate, int bitsPerSample, int channels)
        {
            _sampleRate = sampleRate;
            _bitsPerSample = bitsPerSample;
            _channels = channels;
            _stream = new FileStream(path, FileMode.Create, FileAccess.Write);
            _writer = new BinaryWriter(_stream);
            WriteHeader();
        }

        private void WriteHeader()
        {
            int byteRate = _sampleRate * _channels * (_bitsPerSample / 8);
            short blockAlign = (short)(_channels * (_bitsPerSample / 8));

            // RIFF header
            _writer.Write(new char[] { 'R', 'I', 'F', 'F' });
            _writer.Write(0); // Placeholder for file size - 8
            _writer.Write(new char[] { 'W', 'A', 'V', 'E' });

            // fmt chunk
            _writer.Write(new char[] { 'f', 'm', 't', ' ' });
            _writer.Write(16); // Chunk size
            _writer.Write((short)1); // PCM format
            _writer.Write((short)_channels);
            _writer.Write(_sampleRate);
            _writer.Write(byteRate);
            _writer.Write(blockAlign);
            _writer.Write((short)_bitsPerSample);

            // data chunk
            _writer.Write(new char[] { 'd', 'a', 't', 'a' });
            _dataChunkSizePosition = _stream.Position;
            _writer.Write(0); // Placeholder for data size
        }

        public long Length => _dataSize;

        public void Write(byte[] buffer, int offset, int count)
        {
            if (_disposed) return;
            _writer.Write(buffer, offset, count);
            _dataSize += count;
        }

        public void Flush()
        {
            if (_disposed) return;
            _writer.Flush();
        }

        public void Dispose()
        {
            if (_disposed) return;
            _disposed = true;

            try
            {
                // Update data chunk size
                _stream.Position = _dataChunkSizePosition;
                _writer.Write(_dataSize);

                // Update RIFF chunk size (file size - 8)
                _stream.Position = 4;
                _writer.Write((int)(_stream.Length - 8));

                _writer.Flush();
            }
            catch (Exception) { }

            _writer.Dispose();
            _stream.Dispose();
        }
    }
}

File: htcommander_flutter/lib/radio/wav_file_writer.dart
import 'dart:io';
import 'dart:typed_data';

/// Writes PCM audio data to a WAV file.
///
/// Supports mono 16-bit PCM at any sample rate.
class WavFileWriter {
  final File _file;
  final int sampleRate;
  final int channels;
  final int bitsPerSample;
  RandomAccessFile? _raf;
  int _dataSize = 0;

  WavFileWriter(
    String path, {
    this.sampleRate = 32000,
    this.channels = 1,
    this.bitsPerSample = 16,
  }) : _file = File(path);

  /// Opens the file and writes the WAV header placeholder.
  void open() {
    _raf = _file.openSync(mode: FileMode.write);
    _dataSize = 0;
    _writeHeader();
  }

  /// Writes PCM samples to the file.
  void writeSamples(Uint8List pcmData) {
    final raf = _raf;
    if (raf == null) return;
    raf.writeFromSync(pcmData);
    _dataSize += pcmData.length;
  }

  /// Writes Int16List samples to the file as little-endian bytes.
  void writeInt16Samples(Int16List samples) {
    writeSamples(Uint8List.view(samples.buffer, samples.offsetInBytes,
        samples.lengthInBytes));
  }

  /// Finalizes the WAV file by updating the header with actual sizes.
  void close() {
    final raf = _raf;
    if (raf == null) return;

    // Update RIFF chunk size
    raf.setPositionSync(4);
    _writeUint32LE(raf, 36 + _dataSize);

    // Update data chunk size
    raf.setPositionSync(40);
    _writeUint32LE(raf, _dataSize);

    raf.closeSync();
    _raf = null;
  }

  /// Returns the path of the WAV file.
  String get path => _file.path;

  /// Returns the number of data bytes written so far.
  int get dataSize => _dataSize;

  /// Returns the duration in seconds of audio written so far.
  double get durationSeconds {
    final byteRate = sampleRate * channels * (bitsPerSample ~/ 8);
    if (byteRate == 0) return 0;
    return _dataSize / byteRate;
  }

  void _writeHeader() {
    final raf = _raf!;
    final byteRate = sampleRate * channels * (bitsPerSample ~/ 8);
    final blockAlign = channels * (bitsPerSample ~/ 8);

    // RIFF header
    raf.writeStringSync('RIFF');
    _writeUint32LE(raf, 36); // placeholder, updated on close
    raf.writeStringSync('WAVE');

    // fmt chunk
    raf.writeStringSync('fmt ');
    _writeUint32LE(raf, 16); // chunk size
    _writeUint16LE(raf, 1); // PCM format
    _writeUint16LE(raf, channels);
    _writeUint32LE(raf, sampleRate);
    _writeUint32LE(raf, byteRate);
    _writeUint16LE(raf, blockAlign);
    _writeUint16LE(raf, bitsPerSample);

    // data chunk
    raf.writeStringSync('data');
    _writeUint32LE(raf, 0); // placeholder, updated on close
  }

  static void _writeUint16LE(RandomAccessFile raf, int value) {
    raf.writeByteSync(value & 0xFF);
    raf.writeByteSync((value >> 8) & 0xFF);
  }

  static void _writeUint32LE(RandomAccessFile raf, int value) {
    raf.writeByteSync(value & 0xFF);
    raf.writeByteSync((value >> 8) & 0xFF);
    raf.writeByteSync((value >> 16) & 0xFF);
    raf.writeByteSync((value >> 24) & 0xFF);
  }
}
